cmd/image-builder/upload_test.go

```diff
@@ -46,49 +48,79 @@ func (fa *fakeAwsUploader) UploadAndRegister(r io.Reader, status io.Writer) erro
 func TestUploadWithAWSMock(t *testing.T) {
 	fakeDiskContent := "fake-raw-img"
 
-	fakeImageFilePath := filepath.Join(t.TempDir(), "disk.raw")
-	err := os.WriteFile(fakeImageFilePath, []byte(fakeDiskContent), 0644)
-	assert.NoError(t, err)
-
 	var regionName, bucketName, amiName string
-	var fa fakeAwsUploader
-	restore := main.MockAwscloudNewUploader(func(region string, bucket string, ami string, opts *awscloud.UploaderOptions) (cloud.Uploader, error) {
-		regionName = region
-		bucketName = bucket
-		amiName = ami
-		return &fa, nil
-	})
-	defer restore()
+	var uploadOpts *awscloud.UploaderOptions
 
-	var fakeStdout bytes.Buffer
-	restore = main.MockOsStdout(&fakeStdout)
-	defer restore()
+	for _, tc := range []struct {
+		fakeDiskName       string
+		targetArchArg      string
+		expectedUploadArch string
+		expectedWarning    string
+	}{
+		// simple case: explicit target arch, no warning
+		{"fake-disk.img", "aarch64", "aarch64", ""},
+		// no target arch, detectable from filename: add note
+		{"centos-9-ami-aarch64.img", "", "aarch64", `Note: using architecture "aarch64" based on image filename (use --arch to override)` + "\n"},
+		// no target arch, not detectable form filename: we warn and expect host arch
+		{"fake-disk.img", "", arch.Current().String(), `WARNING: no upload architecture specified, using "x86_64" (use --arch to override)` + "\n"},
```

Comment:
The error message actually interpolates the current architecture into the string, our tests don't only run on `x86_64` so this fails on every other architure.